src/Engine/GameAction.java

```diff
@@ -480,77 +486,90 @@ public String toString()
+    {
+      actor = transport;
+      movePath = path;
+      myDropoffs = dropoffs;
+
+      // Grab the move location and the first drop location to support getMoveLocation and getTargetLocation.
+      if( (null != movePath) && (movePath.getPathLength() > 0 ))
+      {
+        moveLoc = new XYCoord(movePath.getEnd().x, movePath.getEnd().y);
+      }
+      if( !myDropoffs.isEmpty() )
+      {
+        for( XYCoord coord : myDropoffs.values() )
+        {
+          firstDropLoc = coord;
+          break;
+        }
+      }
+    }
+
+    @Override
+    public GameEventQueue getEvents(MapMaster gameMap)
     {
       // UNLOAD actions consist of
       //   MOVE (transport)
       //   UNLOAD
-      actor = transport;
-      unloadEvents = new GameEventQueue();
-      myDropoffs = new HashMap<Unit, XYCoord>();
+      //   [UNLOAD]*
+      GameEventQueue unloadEvents = new GameEventQueue();
 
       // Validate input.
       boolean isValid = true;
-      isValid &= null != transport && !transport.isTurnOver;
-      isValid &= null != dropoffs && !dropoffs.isEmpty();
+      isValid &= null != actor && !actor.isTurnOver;
+      isValid &= null != myDropoffs && !myDropoffs.isEmpty();
       isValid &= movePath.getPathLength() > 0;
       isValid &= null != gameMap;
       if( isValid )
       {
-        isValid &= !transport.heldUnits.isEmpty();
-        moveLoc = new XYCoord(movePath.getEnd().x, movePath.getEnd().y);
-        isValid &= gameMap.isLocationEmpty(transport, moveLoc); // Move location is unoccupied.
-        myDropoffs.putAll(dropoffs);
-        for( Unit unit : myDropoffs.keySet() )
-        {
-          isValid &= gameMap.isLocationEmpty(transport, myDropoffs.get(unit)); // Drop locations are unoccupied.
-          firstDropLoc = (null == firstDropLoc)? myDropoffs.get(unit) : firstDropLoc;
-        }
+        isValid &= !actor.heldUnits.isEmpty();
       }
 
       // Generate events.
       if( isValid )
       {
-        // Move transport to the target location.
-        unloadEvents.add(new MoveEvent(transport, movePath));
-
-        // Debark the passengers.
-        for( Unit unit : myDropoffs.keySet() )
+        // Attempt to move the transport to the target location.
+        if( Utils.enqueueMoveEvent(gameMap, actor, movePath, unloadEvents) )
         {
-          unloadEvents.add(new UnloadEvent(transport, unit, myDropoffs.get(unit)));
+          // Debark the passengers. If a unit can't land, stop unloading.
+          for( Unit unit : myDropoffs.keySet() )
+          {
+            XYCoord dropXY = myDropoffs.get(unit);
+            if( gameMap.isLocationEmpty(actor, dropXY) )
+            {
+              unloadEvents.add(new UnloadEvent(actor, unit, myDropoffs.get(unit)));
+            }
+            else
+            {
+              // Hah, last guy got stuck in the chute. Everyone else is trapped behind them still. Bummer.
+              break;
```

Comment:
Did we ever decide definitively how to handle the case where an attempt to drop in fog is interrupted? This still has the "subsequent drops fail too" approach (which is still not properly ordered).

Reply:
Uhhh.
I *think* it's the actual order, though that's harder to test.
Either way, executing all valid drops is 100% kosher and easy IMO.

Reply:
Done in FogMod2Mod